---
title: "RNAseqHW"
author: "Manpreet S. Katari"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE)
library('dplyr')
library('stringr')
```

### Data source

The data was downloaded from NCBI GEO ([GSE124548](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE124548)). The study looks at the affect of a drug (Lumacaftor/Ivacaftor) to treat cystic fibrosis ( CF ). This drug has been approved for individuals that are homozygous for CFTR ( Cystic fibrosis transmembrane conductance regulator ) mutation. Clinical studies have observed that there is a large variation in the response thus the researchers are looking at RNA expressions to help identify the cause of this variation. Blood samples were taken from : 
- 20 healthy patients ( do not have CF )
- 20 patients that CF ( before treatment )
- 20 patients that have CF ( after treatment - these are paired with those before treatment)

I have extracted the raw read counts from the original matrix and provided it with the homework. The healthy patients have "HC" in their names, CF patients before treatment are labeled "Base" and after treatment are labeled "V2".

For the homework we will compare the HC with Base to do a simple unpaired differentially expression.

### Step 1 (5pts)

Load the file **GSE124548.raw.txt** and create a new dataframe with just the columns with the raw counts for healthy (HC) and CF patients before treatment (Base) and call it **readcount**. Use the *third* column (EntrezID) in the original file as the rownames of readcount.

```{r readdata}
df = read.delim('GSE124548.raw.fixed.txt', sep = '\t')
readcount = data.frame(select(df,contains("HC"),contains("Base")), row.names = df$EntrezID)
#

```


### Step 2 (5pts)

Create a dataframe, called **expgroup**, with one column, labeled **condition**, that correctly assigns which column is healthy and which is CF. Use the column names of readount as rownames of expgroup.

```{r createfactor}
colnames = colnames(readcount)
condition_vec = c()
for (line in colnames)
{
  if (str_detect(line,'HC'))
  {
    condition_vec = append(condition_vec,'Healthy')
  }
  else
  {
    condition_vec = append(condition_vec,'CF')
  }
}
expgroup = data.frame(condition_vec,row.names = colnames)
```

### Step 3 (5pts)

Load the Deseq2 package ( install if necessary ) and create a Counts Dataset Matrix using the command **DESeqDataSetFromMatrix()** and call it **cds**.

```{r runDESeq2} 
library('DESeq2')
cds = DESeqDataSetFromMatrix(design= ~ condition_vec, countData = readcount, colData = expgroup)


#dds <- DESeqDataSetFromMatrix(countData = data, colData = coldata, design= ~ genotype + treatment + genotype:treatment)


```

### Step 4 (5pts)

Use the functions **estimateSizeFactors** and **estimateDispersions** so that Deseq2 can correct for size of library and estimates the dispersion. Plot the dispersion using **plotDispEsts**. What does the graph tell you?



```{r estimate}
cds = estimateSizeFactors(cds)
cds = estimateDispersions(cds)
plotDispEsts(cds)
#

```

### Step 5 (5pts)

Perform the Differential expression and obtain the results using **DESeq** and **results** functions.


```{r deseqres}
cds = DESeq(cds)
resultsNames(cds)
results = results(cds, name = 'condition_vec_Healthy_vs_CF')



```

### Step 6 (5pts)

How many genes have an adjusted p-value of less than 0.05 and log2FoldChange greater than 1 or less than -1 ? Save this list of genes as **diffexpgenes**

```{r diffexp}
dim(subset(results, padj < 0.05 & log2FoldChange > 1))
dim(subset(results, padj < 0.05 & log2FoldChange < -1))
# 17 genes have an adjusted p-value of less than 0.05 and log2FoldChange greater than 1
# 192 genes have an adjusted p-value of less than 0.05 and log2FoldChange less than -1
diffexpgenes = subset(results, padj < 0.05 & (log2FoldChange > 1 | log2FoldChange < -1))

```


### Step 7 (5pts)

Get the normalized values of the counts data in cds using the counts() function with
option normalized=T and call this normvalues.

```{r normvalues}
normvalues = counts(cds,normalized=TRUE)
#


```

### Step 8 (5pts)

Create a new matrix or dataframe that contains the expression values from normvalues for just
the diffexpgenes and call it diffexpvalues.

```{r diffvalues}
diffexpgenes_df = as.data.frame(diffexpgenes)
diffexpvalues = normvalues[rownames(normvalues) %in% rownames(diffexpgenes_df), ] 

#


```

### Step 9 (10pts)

Cluster the differentially expressed genes using hierarchical clustering and use the cutree
function to create 8 groups. How many genes are in each group?

```{r diffgroups}
dist = dist(diffexpgenes_df)
tree = hclust(dist)
clusters = cutree(tree,k = 8)
clusters_df = as.data.frame(clusters)

group_by(clusters_df, clusters) %>% summarize(count=n()) 
#


```

### Step 10 (10pts)

Create a heatmap of all the genes using the **pheatmap** package. Some options you should pay attention to are:
- scale ( scaling gene values shows expression pattern changes rather than magnitude)
- cluster_rows ( define how the rows should be clustered)
- annotation_col ( color code the different samples based on a category- health or cf)
- annotation_row ( color code the different genes based on the cluster they belong to)

```{r heatmap}
library('pheatmap')
pheatmap(readcount, scale = 'row', cluster_rows = TRUE, annotation_col= expgroup, annotation_row= clusters_df)



```


### Step 11 (10pts)

Use the GOstats package to determing which GO-terms are enriched in **diffexpgenes**. To do this you will need to install the following packages from Bioconductor:

```{r eval=F}
BiocManager::install("GOstats")
BiocManager::install("GO.db")
BiocManager::install("Category")
BiocManager::install("org.Hs.eg.db")


```


```{r loadlibraries}
library('GOstats')
library('Category')
library("org.Hs.eg.db")
library('annotate')
library('GO.db')


```

Now create a new **GOHyperGParams** object using the **new()** function and save it as variable called **params**. The **geneIds** is **diffexpgenes**. These should be the EntrezIDs we made into rownames in the beginning. The **universeGeneIds** is the rownames of **readcount**. The annotation is **org.Hs.eg** and the ontology we are going to use is **BP** with a pvaluecutoff of 0.001 and our testDirection is **over**.

```{r gostats}
params = new('GOHyperGParams',geneIds = rownames(diffexpgenes_df), universeGeneIds = rownames(readcount), annotation = "org.Hs.eg.db", ontology = 'BP', pvalueCutoff = 0.001, testDirection = 'over')



```

### Step 12 (5pts)

What conclusions can you make about your analysis?

```

#

```